plugins/inputs/cisco_telemetry_mdt_netconf/cisco_telemetry_mdt_netconf.go

```diff
@@ -0,0 +1,930 @@
+						err,
+					))
+				} else if err == nil {
+					// Take the alarm content only, without the <eventTime> entry (child 0)
+					// Send data to Influx accumulator
+					c.handleTelemetry(tt.Children[1], event.EventTime)
+				}
+			}
+		}
+	}
+	dsrs.state = ended
+	c.Log.Debugf(
+		"Stopped Cisco NETCONF event notification subscription service on %s",
+		c.ServerAddress,
+	)
+}
+
+// createService calls functionalities for the get service
+func (grs *getRequestsService) createService(
+	ctx context.Context, c *CiscoTelemetryNETCONF) {
+	grs.setting = &setting{Client: nil, Session: nil,
+		Mutex: new(sync.Mutex), state: 0}
+
+	defer c.waitgroup.Done()
+
+	for ctx.Err() == nil {
+		c.connectClient(ctx, grs.setting)
+		if grs.setting.state == connected {
+			grs.createRequests(ctx, c)
+		}
+
+		select {
+		case <-ctx.Done():
+		case <-time.After(time.Duration(c.Redial)):
+		}
+	}
+}
+
+// createRequests makes get requests over NETCONF, at a given polling interval
+func (grs *getRequestsService) createRequests(
+	ctx context.Context, c *CiscoTelemetryNETCONF) error {
+	var err error
+	// Waitgroup for goroutines allocated to each get request
+	var waitgroup sync.WaitGroup
+	defer waitgroup.Wait()
+
+	if len(grs.Gets) > 0 {
+
+		// Initialization of get operations
+		gets := make([]*netconf.Get, len(grs.Gets))
+
+		for i, g := range grs.Gets {
+			// Select filter
+			filter := netconf.Filter{
+				Type:   "xpath",
+				Select: g.SelectFilter,
+			}
+
+			gets[i] = &netconf.Get{
+				Filter: &filter,
+			}
+
+			// Add the next goroutine to the waitlist
+			waitgroup.Add(1)
+
+			go func(get *netconf.Get, period config.Duration) {
+				// Create a ticker that ticks periodically until it is stopped
+				ticker := time.NewTicker(time.Duration(
+					time.Duration(period).Nanoseconds()) * time.Nanosecond)
+
+				for ctx.Err() == nil {
+					grs.Mutex.Lock()
+					sessionExists := grs.Session != nil
+					grs.Mutex.Unlock()
+
+					if sessionExists {
+						var err error
+						getReply := new(netconf.RPCReplyData)
+
+						grs.Mutex.Lock()
+						sessionExists := grs.Session != nil
+						if sessionExists {
+							// Get operation
+							c.Log.Debugf(
+								"Performing <get> operation ...",
+							)
+
+							err = grs.Session.Call(get, getReply)
+							grs.Mutex.Unlock()
+
+							// Get the current timestamp because
+							// no timestamp is delivered with NETCONF get
+							timestamp := time.Now()
+
+							if err != nil {
+								c.reportError(ctx, err, "<get> operation")
+							} else {
+								// Unmarshal XML data
+								var tt TelemetryTree
+								err = xml.Unmarshal(getReply.Data.InnerXML, &tt)
+
+								if err != nil {
+									c.acc.AddError(fmt.Errorf(
+										"failed to unmarshal XML: %s",
+										err))
+								} else if err == nil {
+									// Send data to Influx accumulator
+									c.handleTelemetry(tt, timestamp)
+								}
+							}
+						} else {
+							grs.Mutex.Unlock()
+						}
```

Comment:
Why do you need to hold the mutex for the whole time. To me this looks bogus. Please hold the mutex _only_ when changing the relevant field and not across multiple, potentially failing, operations!